packages/cli/src/sso.ee/oidc/oidc.service.ee.ts

```diff
@@ -0,0 +1,257 @@
+import type { OidcConfigDto } from '@n8n/api-types';
+import { Logger } from '@n8n/backend-common';
+import { GlobalConfig } from '@n8n/config';
+import { AuthIdentityRepository, SettingsRepository, type User, UserRepository } from '@n8n/db';
+import { Service } from '@n8n/di';
+import { Cipher } from 'n8n-core';
+import { jsonParse } from 'n8n-workflow';
+import * as client from 'openid-client';
+
+import config from '@/config';
+import { BadRequestError } from '@/errors/response-errors/bad-request.error';
+import { ForbiddenError } from '@/errors/response-errors/forbidden.error';
+import { InternalServerError } from '@/errors/response-errors/internal-server.error';
+import { UrlService } from '@/services/url.service';
+
+import {
+	OIDC_CLIENT_SECRET_REDACTED_VALUE,
+	OIDC_LOGIN_ENABLED,
+	OIDC_PREFERENCES_DB_KEY,
+} from './constants';
+import {
+	getCurrentAuthenticationMethod,
+	isEmailCurrentAuthenticationMethod,
+	isOidcCurrentAuthenticationMethod,
+	setCurrentAuthenticationMethod,
+} from '../sso-helpers';
+
+const DEFAULT_OIDC_CONFIG: OidcConfigDto = {
+	clientId: '',
+	clientSecret: '',
+	discoveryEndpoint: '',
+	loginEnabled: false,
+};
+
+type OidcRuntimeConfig = Pick<OidcConfigDto, 'clientId' | 'clientSecret' | 'loginEnabled'> & {
+	discoveryEndpoint: URL;
+};
+
+const DEFAULT_OIDC_RUNTIME_CONFIG: OidcRuntimeConfig = {
+	...DEFAULT_OIDC_CONFIG,
+	discoveryEndpoint: new URL('http://n8n.io/not-set'),
+};
+
+@Service()
+export class OidcService {
+	private oidcConfig: OidcRuntimeConfig = DEFAULT_OIDC_RUNTIME_CONFIG;
+
+	constructor(
+		private readonly settingsRepository: SettingsRepository,
+		private readonly authIdentityRepository: AuthIdentityRepository,
+		private readonly urlService: UrlService,
+		private readonly globalConfig: GlobalConfig,
+		private readonly userRepository: UserRepository,
+		private readonly cipher: Cipher,
+		private readonly logger: Logger,
+	) {}
+
+	async init() {
+		this.oidcConfig = await this.loadConfig(true);
+		await this.setOidcLoginEnabled(this.oidcConfig.loginEnabled);
+	}
+
+	getCallbackUrl(): string {
+		return `${this.urlService.getInstanceBaseUrl()}/${this.globalConfig.endpoints.rest}/sso/oidc/callback`;
+	}
+
+	getRedactedConfig(): OidcConfigDto {
+		return {
+			...this.oidcConfig,
+			discoveryEndpoint: this.oidcConfig.discoveryEndpoint.toString(),
+			clientSecret: OIDC_CLIENT_SECRET_REDACTED_VALUE,
+		};
+	}
+
+	async generateLoginUrl(): Promise<URL> {
+		const configuration = await this.getOidcConfiguration();
+
+		const authorizationURL = client.buildAuthorizationUrl(configuration, {
```

Comment:
2 things here:
- It seems we do not support PKCE here, I'm not sure of how critical it is, but I'm wondering if maybe some providers enforces it
- Maybe the scope should be configurable, no ? I'm thinking of organisation which do not want to enable profile for instance - not sure if critical

Reply:
Both things are correct, but we want to have some OIDC support as fast as possible. I already told marc and jc that we should add support for PKCE. Making the scope configurable, would also be an good option.